algorithms/estonian/stemmer.tt

```diff
@@ -0,0 +1,479 @@
+'hüppaksite',
+'hüppan',
+'hüppad',
+'hüppab',
+'hüppame',
+'hüppate',
+'hüppavad',
+'hüppasin',
+'hüppasid',
+'hüppas',
+'hüppasime',
+'hüppasite',
+'hüpanuksite',
+'hüpatakse',
+'hüpati',
+'hüpanud',
+'hüpanutest',
+'hüpates',
+'hüppavat',
+'hüppavatele',
+'hüppamata',
+'hüppamast',
+'hüljes',
+'hülge',
+'hüljest',
+'hülgesse',
+'hüljeste',
+'hülgeid',
+'hüljestesse',
+'hülgeisse',
+'ohutule',
+'ohutud',
+'ohutuid',
+'ohututele',
+]) %]
+</div>
+
+<p>
+This algorithm is written in collaboration with Estonian text analytics enterprise Texta.
+</p>
+
+<p>
+Letters in Estonian include the following accented forms,
+</p>
+
+<DL><DD>
+    <B><I>ä  &nbsp;  ö  &nbsp;  õ  &nbsp;  ü</I></B>
+</DL>
+
+<p>
+The following letters are vowels (<I>V1</I>):
+</p>
+
+<DL><DD>
+    <B><I>a  &nbsp;  e  &nbsp;  i  &nbsp;  o  &nbsp;  u  &nbsp;  õ  &nbsp;  ä  &nbsp;  ö  &nbsp;  ü</I></B>
+</DL>
+
+<p>
+<I>RV</I> is defined as one of the following:
+</p>
+
+<DL><DD>
+    <B><I>a  &nbsp;  e  &nbsp;  i  &nbsp;  u  &nbsp;  o</I></B>
+</DL>
+
+<p>
+<I>KI</I> is defined as one of the following (letters possible before <B><I>-ki</I></B> emphasis):
+</p>
+
+<DL><DD>
+    <B><I>k  &nbsp;  p  &nbsp;  t  &nbsp;  g  &nbsp;  b  &nbsp;  d  &nbsp;  s  &nbsp;  h  &nbsp;  f  &nbsp;  š  &nbsp;  z  &nbsp;  ž  &nbsp;  o</I></B>
+</DL>
+
+<p>
+<I>GI</I> is defined as one of the following (letters possible before <B><I>-gi</I></B> emphasis):
+</p>
+
+<DL><DD>
+    <B><I>c  &nbsp;  j  &nbsp;  l  &nbsp;  m  &nbsp;  n  &nbsp;  q  &nbsp;  r  &nbsp;  v  &nbsp;  w  &nbsp;  x  &nbsp;  a  &nbsp;  e  &nbsp;  i  &nbsp;  o  &nbsp;  u  &nbsp;  õ  &nbsp;  ä  &nbsp;  ö  &nbsp;  ü</I></B>
+</DL>
+
+<p>
+<I>R1</I> in this algorithm is defined as a region after the first consonant preceded by a vowel ([laul[nud], mõt[teid], kar[tuleid], saab[as] ). If there’s no such region, then the <I>R</I> is empty (laul[Ø], saun[Ø]). Limitations in steps (such as ‘preceded by <I>RV</I>’ ) are not restricted to the <I>R1</I> region.
+</p>
+
+<p>
+<I>LONGV</I> is defined as one of the following:
+</p>
+
+<DL><DD>
+    <B><I>aa  &nbsp;  ee  &nbsp;  ii  &nbsp;  oo  &nbsp;  uu  &nbsp;  ää  &nbsp;  öö  &nbsp;  üü  &nbsp;  õõ</I></B>
+</DL>
+
+<p>
+Do step 0. If nothing was changed in step 0, continue with the steps, otherwise stop. Do step 1 and step 2. If nothing was changed in step 2, do steps 3, 4, 5, 6, 7 , 8 and 9. If something was changed in  step 2, do step 9.
+</p>
+
+<p>
+Step 0: verb_exceptions
+</p>
+
+<DL><DD>
+    Search for some frequent irregular short verbs which wouldn’t have been found otherwise and give them a chosen stem.
+<BR><BR>
+<DL>
+    <DT><B><I>joon  &nbsp;  jood  &nbsp;  joob  &nbsp;  joote  &nbsp;  joome  &nbsp;  joovad</I></B>
+        <DD>replace by <B><I>joo</I></B>
+    <DT><B><I>jõin  &nbsp;  jõid  &nbsp;  jõi  &nbsp;  jõime  &nbsp;  jõite</I></B>
+        <DD>replace by <B><I>joo</I></B>
+    <DT><B><I>juuakse  &nbsp;  joodakse  &nbsp;  juua  &nbsp;  jooma  &nbsp;  joomata</I></B>
```

Comment:
The words here are in a different order to those in the `.sbl` file - while that's not a bug in the code or this description as such, it's really unhelpful for someone trying to compare the two so I think it better avoided.  Which order makes more sense to standardised on?

Reply:
Looking at the declensions in wiktionary, the `.sbl` order seems more logical.